skfda/clustering.py

```diff
@@ -0,0 +1,1180 @@
+                algorithm.
+            random_state (int, RandomState instance or None): Defaults to 0.
+                Determines random number generation for centroid initialization.
+                Use an int to make the randomness deterministic. See Glossary.
+                https://scikit-learn.org/stable/glossary.html#term-random-state
+            p (int): Identifies the p-norm used to calculate the distance
+                between functions.
+        """
+        if max_iter < 1:
+            raise ValueError(
+                "The number of iterations must be greater than 0.")
+
+        self._max_iter = max_iter
+        self._random_state = random_state
+        self._p = p
+
+    @property
+    def fdatagrid(self):
+        return self._fdatagrid
+
+    @property
+    def n_clusters(self):
+        return self._n_clusters
+
+    @property
+    def init(self):
+        return self._init
+
+    @property
+    def max_iter(self):
+        return self._max_iter
+
+    @property
+    def random_state(self):
+        return self._random_state
+
+    @property
+    def p(self):
+        return self._p
+
+    @property
+    def centers(self):
+        return self._centers
+
+    @property
+    def n_iter(self):
+        return self._n_iter
+
+    def _generic_clustering_checks(self, fdatagrid, n_clusters, init):
+        """Checks the arguments *fdatagrid*, *n_clusters* and *init*
+        used in the :func:`fit <fda.clustering.ClusteringData.fit>` function.
+
+        Args:
+            fdatagrid (FDataGrid object): Object whose samples
+                are classified into different groups.
+            n_clusters (int): Number of groups into which the samples are
+                classified.
+            init (ndarray): Contains the initial centers of the different
+                clusters the algorithm starts with or None.
+
+        Returns:
+            fdatagrid (FDataGrid object): Object whose samples
+                are classified into different groups.
+            n_clusters (int): Number of groups into which the samples are
+                classified.
+            init (ndarray): Contains the initial centers of the different
+                clusters the algorithm starts with. If None, default value,
+                the centers are generated randomly.
+        """
+
+        if fdatagrid.ndim_domain > 1:
+            raise NotImplementedError(
+                "Only support 1 dimension on the domain.")
+
+        if fdatagrid.nsamples < 2:
+            raise ValueError(
+                "The number of observations must be greater than 1.")
+
+        if n_clusters < 2:
+            raise ValueError(
+                "The number of clusters must be greater than 1.")
+
+        if init is not None and init.shape != (
+                fdatagrid.ndim_image, n_clusters, fdatagrid.ncol):
+            raise ValueError(
+                "The init ndarray should be of shape (ndim_image, n_clusters, n_features) "
```

Comment:
Why this shape? I would expect that a fdatagrid is received, with as many observations as clusters.

Reply:
An array with as many observations as clusters is expected, but ndim_image times. There should be the possibility to initialize different centroids for each dimension of the image.
I could change it to expect an fdatagrid, in that case, its data_matrix should be of the shape (n_clusters, n_features, ndim_image).

Reply:
I think a FDataGrid is better. Maybe you can allow an automatic conversion from an array, but the shape should be the same as the FDataGrid.